flowmachine/flowmachine/core/cache.py

```diff
@@ -0,0 +1,384 @@
+
+# -*- coding: utf-8 -*-
+
+"""
+Functions which deal with inspecting and managing the query cache.
+"""
+import logging
+import pickle
+
+from typing import TYPE_CHECKING, Tuple, List
+
+from psycopg2 import InternalError
+
+if TYPE_CHECKING:
+    from .query import Query
+    from .connection import Connection
+
+logger = logging.getLogger("flowmachine").getChild(__name__)
+
+
+def touch_cache(connection: "Connection", query_id: str) -> float:
+    """
+    'Touch' a cache record and update the cache score for it.
+
+    Parameters
+    ----------
+    connection : Connection
+    query_id : str
+        md5 id of the query to touch
+
+    Returns
+    -------
+    float
+        The new cache score
+    """
+    try:
+        return float(connection.fetch(f"SELECT touch_cache('{query_id}')")[0][0])
+    except (IndexError, InternalError):
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def reset_cache(connection: "Connection") -> None:
+    """
+    Reset the query cache. Deletes any tables under cache schema, resets the cache count
+    and clears the cached and dependencies tables.
+
+    Parameters
+    ----------
+    connection : Connection
+    """
+    qry = f"SELECT tablename FROM cache.cached WHERE NOT class='Table'"
+    tables = connection.fetch(qry)
+    with connection.engine.begin() as trans:
+        trans.execute("SELECT setval('cache.cache_touches', 1)")
+        for table in tables:
+            trans.execute(f"DROP TABLE IF EXISTS cache.{table[0]} CASCADE")
+        trans.execute("TRUNCATE cache.cached CASCADE")
+        trans.execute("TRUNCATE cache.dependencies CASCADE")
+
+
+def invalidate_cache_by_id(
+    connection: "Connection", query_id: str, cascade=False
+) -> "Query":
+    """
+    Remove a query object from cache by id.
+
+    Parameters
+    ----------
+    connection : Connection
+    query_id : str
+        md5 id of the query
+    cascade : bool, default False
+        Set to true to remove any queries that depend on the one being removed
+
+    Returns
+    -------
+    Query
+        The original query object.
+
+    """
+    query_obj = get_query_by_id(connection, query_id)
+    query_obj.invalidate_db_cache(cascade=cascade)
+    return query_obj
+
+
+def get_query_by_id(connection: "Connection", query_id: str) -> "Query":
+    """
+    Get a query object from cache by id.
+
+    Parameters
+    ----------
+    connection : Connection
+    query_id : str
+        md5 id of the query
+
+    Returns
+    -------
+    Query
+        The original query object.
+
+    """
+    qry = f"SELECT obj FROM cache.cached WHERE query_id='{query_id}'"
+    try:
+        obj = connection.fetch(qry)[0][0]
+        return pickle.loads(obj)
+    except IndexError:
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def get_cached_queries_by_score(connection: "Connection") -> List[Tuple["Query", int]]:
```

Comment:
See above: rename this?
```suggestion
def get_cached_query_objects_by_score(connection: "Connection") -> List[Tuple["Query", int]]:
```